### ISSUE TITLE
Certificate Management

### ISSUE THREAD

**Issue by MJRichardson** [OctopusDeploy/Issues#2701]
## What

Allow storage and management of X509 Certificates within Octopus
## Why

The impetus for this feature came from our goal to provide the best possible experience for deploying IIS websites.  Currently, a significant pain-point is managing the digital certificates used for HTTPS bindings.  The certificates must currently be referenced by thumbprint, and must be already installed on the machine.  Installing the certificates by custom scripts is not trivial.

The idea is the allow uploading certificates into Octopus, and reference them when configuring HTTPS bindings.  
## Taking it further

Management of certificates is a problem for many organisations, and Octopus seems like a natural fit to implement a solution.

In addition to referencing managed certificate when creating IIS bindings, other potential features include:
- Allow using managed certificates in custom scripts (and make it easy to do so)
- Give notifications when certificate are close to expiry
- Report which machines certificates are installed on
- Automate replacing certificates with a new certificate
- Provide integration with [Let's Encrypt](https://letsencrypt.org/) 
### The Holy Grail

The ultimate endgame in our minds would be something like:
- Add a new certificate (automatically request from Let's Encrypt).
- Reference it in HTTPS binding when creating IIS deployment step
- When certificate is close to expiry, new certificate is automatically requested from Let's Encrypt and automatically replace previous certificate on all machines where it was installed.

In other words: configure it once, then never think about it again.
## How

The plan is to perform this work in phases:
- [x] [Phase 1](https://github.com/OctopusDeploy/Issues/issues/2702): Allow uploading certificates into Octopus and referencing them from HTTPS binding configuration and using them in custom scripts.
- [ ] Phase 2:  Display which machines certificates are installed on, and allow automatic replacement.
- [ ] Phase 3: Let's Encrypt integration.

**Comment by MJRichardson** [comment]
Phase 2 and 3 will be raised as separate issues if and when they are scheduled for implementation.

**Comment by jballe** [comment]
Any news on phase 2 and 3 since Let's encrypt support for Octopus Server itself has been released?

**Comment by SimonLuckenuik** [comment]
@MJRichardson , any update for phase 2 and 3? I am interested in Let's Encrypt support!
